File: cpp/Ch 6. Math and Logic Puzzles/Intro/SieveOfEratosthenes.cpp
#include <string>
#include <vector>
#include <iostream>
#include <cmath>

void crossOff(std::vector<bool> &flags, int prime)
{		 	
		for (int i = prime * prime; i < flags.size(); i += prime)
		{
			flags[i] = false;
		}
}

	int getNextPrime(std::vector<bool> &flags, int prime)
	{
		int next = prime + 1;
		while (next < flags.size() && !flags[next])
		{
			next++;
		}
		return next;
	}

	void init(std::vector<bool> &flags)
	{
		flags[0] = false;
		flags[1] = false;
		for (int i = 2; i < flags.size(); i++)
		{
			flags[i] = true;
		}
	}

	void std::vector<int> prune(std::vector<bool> &flags, int count)
	{
		std::vector<int> primes(count);
		int index = 0;
		for (int i = 0; i < flags.size(); i++)
		{
			if (flags[i])
			{
				primes[index] = i;
				index++;
			}
		}
		return primes;
	}

	void std::vector<bool> sieveOfEratosthenes(int max)
	{
		std::vector<bool> flags(max + 1);

		init(flags);
		int prime = 2;

		while (prime <= sqrt(max))
		{
			   
			crossOff(flags, prime);

			
			prime = getNextPrime(flags, prime);
		}

		return flags;
	}

	void main(std::vector<std::wstring> &args)
	{
		std::vector<bool> primes = sieveOfEratosthenes(4);
		for (int i = 0; i < primes.size(); i++)
		{
			if (primes[i])
			{
				std::wcout << i << std::endl;
			}
		}
	}

}

File: Java/Ch 06. Math and Logic Puzzles/Introduction/SieveOfEratosthenes.java
package Introduction;

public class SieveOfEratosthenes {
	
	public static void crossOff(boolean[] flags, int prime) {
		/* Cross off remaining multiples of prime. We can start with
		 * (prime*prime), because if we have a k * prime, where k < prime,
		 * this value would have already been crossed off in a prior
		 * iteration. */		
		for (int i = prime * prime; i < flags.length; i += prime) {
			flags[i] = false;
		}
	}
	
	public static int getNextPrime(boolean[] flags, int prime) {
		int next = prime + 1;
		while (next < flags.length && !flags[next]) {
			next++;
		}
		return next;	
	}
	
	public static void init(boolean[] flags) {
		flags[0] = false;
		flags[1] = false;
		for (int i = 2; i < flags.length; i++) {
			flags[i] = true;
		}
	}
	
	public static int[] prune(boolean[] flags, int count) {
		int[] primes = new int[count];
		int index = 0;
		for (int i = 0; i < flags.length; i++) {
			if (flags[i]) {
				primes[index] = i;
				index++;
			}
		}
		return primes;
	}
	
	public static boolean[] sieveOfEratosthenes(int max) {
        boolean[] flags = new boolean[max + 1];
        
		init(flags);
        int prime = 2;
        
        while (prime <= Math.sqrt(max)) {       	
    		/* Cross off remaining multiples of prime */   
    		crossOff(flags, prime); 

    		/* Find next value which is true */
    		prime = getNextPrime(flags, prime);
        }
        
        return flags; //prune(flags, count);
	}
	
	public static void main(String[] args) {
		boolean[] primes = sieveOfEratosthenes(4);
		for (int i = 0; i < primes.length; i++) {
			if (primes[i]) {
				System.out.println(i);
			}
		}
	}

}
